gateway-messages/src/lib.rs

```diff
@@ -141,6 +141,72 @@ pub enum SwitchDuration {
     Forever,
 }
 
+/// Sensor readings that we could request from the target by `SensorId`
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, SerializedSize, Serialize, Deserialize,
+)]
+pub enum SensorRequestKind {
+    /// Requests the most recent reading, which is either a value or error
+    LastReading,
+    /// Requests the most recent data value
+    LastData,
+    /// Requests the most recent error value
+    LastError,
+    /// Requests the error count for a given sensor
+    ErrorCount,
+}
+
+/// Sensor readings that we could request from the target by `SensorId`
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, SerializedSize, Serialize, Deserialize,
+)]
+pub struct SensorRequest {
+    pub kind: SensorRequestKind,
+    pub id: u32,
+}
+
+/// Most recent sensor reading, which may be a reading or a value
+#[derive(
+    Debug, Clone, Copy, PartialEq, SerializedSize, Serialize, Deserialize,
+)]
+pub struct SensorReading {
+    pub value: Result<f32, SensorDataMissing>,
+    pub timestamp: u64,
+}
+
+/// Response to a [`SensorRequest`]
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    SerializedSize,
+    Serialize,
+    Deserialize,
+    strum_macros::IntoStaticStr,
+)]
+#[strum(serialize_all = "snake_case")]
+pub enum SensorResponse {
+    LastReading(SensorReading),
+    LastData { value: f32, timestamp: u64 },
```

Comment:
Couple questions here:

1) What is supposed to be teh semantic difference issuing `LastReading` and `LastData`? When would I use one or the other?

2) If there's only a way to get the last reading, does this basically mean that if we're not polling in a hard-real time fashion (that is in the deadline sense of real-time), we're going to lose data in the broader system for any stat that's either discrete or a gauge as opposed to a counter?

3) Out of curiosity, why the choice of normalizing values to floats? I'm guessing that might be what it already was stored as, but it wouldn't have been my first guess / expectation? I assume when we have a discrete sensor like a power good it'll be up to the consumer to know what it represents and translate that into the discrete expected states?

4) As a consumer, it's sometimes helpful to know what the length of time this covers is. For example, we're inserting into Clickhouse and probably telling it that covers a particular range I expect. In particular, for counters, if the data is updating every 5s and we're inserting into Clickhouse every 10s, that makes a difference or vice versa.

5) I assume that the pattern here is that collectors would go back and forth a few times to try to correlate a rolling count in milis with themselves to try and figure out what the relationship between the SP's monotonic time is and the real-world time, which is why you added the alternative way to get the current time. Is that right? I think we'll want to think through what the queuing and retry delays are going to be on how we do this accurately.

@bnaecker can you also take a look at this please to see if this'll give us enough information to actually get things into Clickhouse?

Reply:
> 1. What is supposed to be teh semantic difference issuing LastReading and LastData? When would I use one or the other?

`LastReading` returns the current state of the sensor, which is either a data value or an error state.  `LastData` returns the last known data value (or a top-level `NoData` error).  I suspect we'll always want to use `LastReading`, but wanted to build in some flexibility.

> 2. If there's only a way to get the last reading, does this basically mean that if we're not polling in a hard-real time fashion (that is in the deadline sense of real-time), we're going to lose data in the broader system for any stat that's either discrete or a gauge as opposed to a counter?

That's currently the case – the Hubris `sensors` task only stores a single data point per sensor.  We could change that (at the cost of more RAM), in which case we'd probably add new variants to the `enum ReadSensor`.

> 3. Out of curiosity, why the choice of normalizing values to floats? I'm guessing that might be what it already was stored as, but it wouldn't have been my first guess / expectation? I assume when we have a discrete sensor like a power good it'll be up to the consumer to know what it represents and translate that into the discrete expected states?

This is also based on how stuff is implemented on the Hubris side.  The actual discussion predate me, but IIRC Cliff had strong opinions about storing all sensor data as floats representing real units (versus, for example, raw bit patterns reported by sensors).

> 4. As a consumer, it's sometimes helpful to know what the length of time this covers is. For example, we're inserting into Clickhouse and probably telling it that covers a particular range I expect. In particular, for counters, if the data is updating every 5s and we're inserting into Clickhouse every 10s, that makes a difference or vice versa.

This is an interesting point.  From the POV of the `sensors` task, readings are simply instantaneous samples (without a notion of range).  It's true that sensors are polled at some frequency, but that's determined by tasks that provide data into `sensors` (and isn't an inherent property of the sensors or an individual reading).

> 5. I assume that the pattern here is that collectors would go back and forth a few times to try to correlate a rolling count in milis with themselves to try and figure out what the relationship between the SP's monotonic time is and the real-world time, which is why you added the alternative way to get the current time. Is that right? I think we'll want to think through what the queuing and retry delays are going to be on how we do this accurately.

I'm not sure what pattern we'll want to implement here!  I'm trying to expose mostly-raw data from the SP, and since the data includes timestamps, I figured we'll want _some_ way to decode them into host time; asking the SP about its current time seems like a useful building block.

Reply:
@rmustacc From a first glance, there should be no issues getting this data into ClickHouse. I do have a few thoughts, mostly with an eye towards implementing the path to the database.

Something on the host system (MGS, I'm assuming) would want to poll this data "fast enough" to make sure we don't lose our desired temporal resolution. That will probably vary by sensor. MGS or some other component will need to register with `nexus` as a metric producer, at which point an `oximeter` instance will be assigned to collect data from it. That registration includes the collection interval, which is independent of the interval on which the sensors are polled. I'm just pointing out that whatever produces data must (1) be able to make HTTP request to `nexus` and (2) run an HTTP server itself, that is reachable over the underlay network.

The data will need to be slightly reorganized, but nothing major. First `oximeter` supports `f64` measurements, not `f32`. No big deal there, and we can add that if it's important. Next, data needs to be organized into a _target_, the thing we're collecting data about, and a _metric_, the data point being measured. For example, it might make sense to make the thing the sensor is attached to into the target. What the sensor is measuring, say temperature, is the metric. This might be in the weeds implementation-wise, but `Target` and `Metric` are derivable traits in the `oximeter` crate. There are several examples in the `oximeter-producer` crate, `nexus`, `dendrite`, and `crucible`, and I'm always happy to help get more data into ClickHouse.

Reply:
Thanks for the feedback!  Do you think any of this affects the MGS message types, or can that reorganization be done on the host?

e.g. I'd rather not do `f32 -> f64` conversion on the SP, because I don't think it has hardware support for `f64`'s.

Reply:
I discussed this a bit with @bnaecker and I think the organizational part of Targets/Metrics will be what we get from topo so there's nothing more that we need to do at the MGS layer today (assuming that relying on topo is sufficient, which I guess is a bit of a risk).

Reply:
> > 2. If there's only a way to get the last reading, does this basically mean that if we're not polling in a hard-real time fashion (that is in the deadline sense of real-time), we're going to lose data in the broader system for any stat that's either discrete or a gauge as opposed to a counter?
> 
> That's currently the case – the Hubris `sensors` task only stores a single data point per sensor. We could change that (at the cost of more RAM), in which case we'd probably add new variants to the `enum ReadSensor`.

OK. I think probably the way out of this is to eventually add stats that cover how many times have we seen state X for discrete stats and min/max or some percentile for gauges over time.

> > 4. As a consumer, it's sometimes helpful to know what the length of time this covers is. For example, we're inserting into Clickhouse and probably telling it that covers a particular range I expect. In particular, for counters, if the data is updating every 5s and we're inserting into Clickhouse every 10s, that makes a difference or vice versa.
> 
> This is an interesting point. From the POV of the `sensors` task, readings are simply instantaneous samples (without a notion of range). It's true that sensors are polled at some frequency, but that's determined by tasks that provide data into `sensors` (and isn't an inherent property of the sensors or an individual reading).

I think this becomes most important to think through in the future when we basically have oximeter polling at a different frequency or that it has less real-time fidelity and is trying to poll this source at a different rate. It's probably fine, but just something in my head.

> > 5. I assume that the pattern here is that collectors would go back and forth a few times to try to correlate a rolling count in milis with themselves to try and figure out what the relationship between the SP's monotonic time is and the real-world time, which is why you added the alternative way to get the current time. Is that right? I think we'll want to think through what the queuing and retry delays are going to be on how we do this accurately.
> 
> I'm not sure what pattern we'll want to implement here! I'm trying to expose mostly-raw data from the SP, and since the data includes timestamps, I figured we'll want _some_ way to decode them into host time; asking the SP about its current time seems like a useful building block.

OK, makes sense.